File: main/k-candy-store/k-candy-store.cs
using System;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]

internal static class Arith {
    internal static int[,] ComputeBinomialCoefficientsModulo(int size, int mod)
    {
        if (size < 2) size = 2;

        var c = new int[size, size];
        c[0, 0] = c[1, 0] = c[1, 1] = 1;

        for (var n = 2; n != size; ++n) {
            c[n, 0] = c[n, n] = 1;

            for (var r = 1; r != n; ++r)
                c[n, r] = (c[n - 1, r - 1] + c[n - 1, r]) % mod;
        }

        return c;
    }
}

internal struct Pair<T> {
    public T X, Y;
}

internal static class Solution {
    private static int GetValue() => int.Parse(Console.ReadLine());

    private static void Main()
    {
        const int mod = 1000000000;

        var a = new Pair<int>[GetValue()];
        var max = 1;

        for (var i = 0; i != a.Length; ++i) {
            a[i].X = GetValue();
            a[i].Y = GetValue();
            a[i].X += a[i].Y - 1;
            if (max < a[i].X) max = a[i].X;
        }

        var c = Arith.ComputeBinomialCoefficientsModulo(max + 1, mod);
        Array.ForEach(a, p => Console.WriteLine(c[p.X, p.Y]));
    }
}

File: main/k-candy-store/k-candy-store.java
final class Solution {
    static int[][] computeBinomialCoefficientsModulo(int size, final int mod)
    {
        if (size < 2) size = 2;
        
        final int[][] c = new int[size][size];
        c[0][0] = c[1][0] = c[1][1] = 1;
        
        for (int n = 2; n != size; ++n) {
            c[n][0] = c[n][n] = 1;
            
            for (int r = 1; r != n; ++r)
                c[n][r] = (c[n - 1][r - 1] + c[n - 1][r]) % mod;
        }
        
        return c;
    }
    
    public static void main(final String[] args)
    {
        final int mod = 1000000000;
        
        try (final java.util.Scanner sc = new java.util.Scanner(System.in)) {
            final int t = sc.nextInt();
            final int[] ns = new int[t], ks = new int[t];
            
            int max = 1;
            for (int i = 0; i != t; ++i) {
                ns[i] = sc.nextInt();
                ks[i] = sc.nextInt();
                ns[i] += ks[i] - 1;
                if (max < ns[i]) max = ns[i];
            }
            
            final int[][] c = computeBinomialCoefficientsModulo(max + 1, mod);
            for (int i = 0; i != t; ++i) System.out.println(c[ns[i]][ks[i]]);
        }
    }
}
